clash-prelude/tests/Clash/Tests/Resize.hs

```diff
@@ -26,16 +45,146 @@ indexProp Proxy v =
   else
     discard
 
+maybeResizePropT ::
+  forall f a b .
+  ( Integral (f a)
+  , Bounded (f a)
+  , Show (f a)
+  , Integral (f b)
+  , Bounded (f b)
+  , Show (f b)
+  , Resize.Resize f
+  , KnownNat a
+  , KnownNat b
+  ) =>
+  Proxy (f a) ->
+  Proxy (f b) ->
+  H.PropertyT IO ()
+maybeResizePropT _ _ = do
+  let
+    minFa = fromIntegral (minBound @(f a)) :: Integer
+    maxFa = fromIntegral (maxBound @(f a)) :: Integer
+    minFb = fromIntegral (minBound @(f b)) :: Integer
+    maxFb = fromIntegral (maxBound @(f b)) :: Integer
+
+  H.footnote $ "minFa: " <> show minFa
+  H.footnote $ "maxFa: " <> show maxFa
+  H.footnote $ "minFb: " <> show minFb
+  H.footnote $ "maxFb: " <> show maxFb
+
+  input <- H.forAll $ Gen.integral (Range.linear minFa maxFa)
+  let output = Resize.maybeResize @a @b @f (fromIntegral input)
+  if minFb <= input && input <= maxFb
+  then output === Just (fromIntegral input)
+  else output === Nothing
+
+maybeResizeUnsignedProp :: H.Property
+maybeResizeUnsignedProp = H.property $ do
+  a <- H.forAll $ Gen.integral (Range.linear 0 100)
+  b <- H.forAll $ Gen.integral (Range.linear 0 100)
+  withSomeSNat a $ \(SNat :: SNat a) -> do
+    withSomeSNat b $ \(SNat :: SNat b) -> do
+      maybeResizePropT @Unsigned @a @b Proxy Proxy
+
+maybeResizeSignedProp :: H.Property
+maybeResizeSignedProp = H.property $ do
+  a <- H.forAll $ Gen.integral (Range.linear 0 100)
+  b <- H.forAll $ Gen.integral (Range.linear 0 100)
+  withSomeSNat a $ \(SNat :: SNat a) -> do
+    withSomeSNat b $ \(SNat :: SNat b) -> do
+      maybeResizePropT @Signed @a @b Proxy Proxy
+
+maybeResizeIndexProp :: H.Property
+maybeResizeIndexProp = H.property $ do
+  a <- H.forAll $ Gen.integral (Range.linear 1 100)
+  b <- H.forAll $ Gen.integral (Range.linear 1 100)
+  withSomeSNat a $ \(SNat :: SNat a) -> do
+    withSomeSNat b $ \(SNat :: SNat b) -> do
+      maybeResizePropT @Index @a @b Proxy Proxy
+
+maybeResizeBitVectorProp :: H.Property
+maybeResizeBitVectorProp = H.property $ do
+  a <- H.forAll $ Gen.integral (Range.linear 0 100)
+  b <- H.forAll $ Gen.integral (Range.linear 0 100)
+  withSomeSNat a $ \(SNat :: SNat a) -> do
+    withSomeSNat b $ \(SNat :: SNat b) -> do
+      maybeResizePropT @BitVector @a @b Proxy Proxy
+
+maybeTruncatePropT ::
+  forall f a b .
+  ( Integral (f a)
+  , Bounded (f a)
+  , Show (f a)
+  , Integral (f (a + b))
+  , Bounded (f (a + b))
+  , Show (f (a + b))
+  , Resize.Resize f
+  , KnownNat a
+  , KnownNat b
+  ) =>
+  Proxy (f a) ->
+  Proxy (f (a + b)) ->
+  H.PropertyT IO ()
+maybeTruncatePropT _ _ = do
+  let
+    minFa = fromIntegral (minBound @(f a)) :: Integer
+    maxFa = fromIntegral (maxBound @(f a)) :: Integer
+    minFab = fromIntegral (minBound @(f (a + b))) :: Integer
+    maxFab = fromIntegral (maxBound @(f (a + b))) :: Integer
+
+  H.footnote $ "minFa: " <> show minFa
+  H.footnote $ "maxFa: " <> show maxFa
+  H.footnote $ "minFab: " <> show minFab
+  H.footnote $ "maxFab: " <> show maxFab
```

Comment:
This is a truly fabulous footnote!

(Also, thanks for adding footnotes, it's very nice in the case a counterexample is found)